### PR TITLE
Add udp packet support

### PR DESCRIPTION
**Pull request by jammiess**
Add support for the ip packet type different types of protocols that it encapsulates.

### CODE REVIEW

**Comment by jammiess** [comment]
Make sure to update version number before merging

File: src/ip.rs [hunk 1]
```diff
@@ -410,13 +482,93 @@ impl TcpPacket {
...
+    /// than the length of the slice.
+    pub fn parse_from_bytes(data: &dyn AsRef<[u8]>) -> Result<Self, UdpParseErr> {
+        let data = data.as_ref();
+        if data.len() < 8 {
+            return Err(UdpParseErr::MissingHeader);
+        }
+        let source = ((data[0] as u16) << 8) + (data[1] as u16);
+        let dest = ((data[2] as u16) << 8) + (data[3] as u16);
+        let length = ((data[4] as u16) << 8) + (data[5] as u16);
+        let length = length as usize;
```
Comment (johnyburd) [inline src/ip.rs]: seems like rust should have neater way of flipping these

File: src/replay.rs [hunk 2]
```diff
@@ -89,27 +89,30 @@ impl Arbiter for AddressArbiter {
...
-        if self.host == packet.dest && self.client == packet.source {
-            // Has the right destination and source, now check that actually data was sent
-            if packet.payload.flags.contains(TcpFlags::PSH) {
-                return Action::Send(None);
-            }
-            // Check if connection got reset and break if it did
-            if packet.payload.flags.intersects(CONN_END) {
-                return Action::Break;
+        // If the packet is not a TCP packet just skip it
+        if let DataType::TCP(ref payload) = packet.payload {
```
Comment (johnyburd) [inline src/replay.rs:93]: no support for udp in the replayer?

File: src/ip.rs [hunk 3]
```diff
@@ -410,13 +482,93 @@ impl TcpPacket {
...
+
+impl UdpPacket {
+    /// Parse a udp packet from a slice of bytes
+    ///
+    /// This method will not check the checksum.
+    ///
+    /// # Errors
+    /// This method will return an error if the slice is less than 8 bytes or if the size of the length field is larger
+    /// than the length of the slice.
+    pub fn parse_from_bytes(data: &dyn AsRef<[u8]>) -> Result<Self, UdpParseErr> {
```
Comment (johnyburd) [inline src/ip.rs:520]: this function looks right to me

File: src/ip.rs [hunk 4]
```diff
@@ -410,13 +482,93 @@ impl TcpPacket {
...
+    /// Size in header is too large for passed in slice
+    #[error("Size in header did not match size of data")]
+    SizeMismatch,
+    /// The slice is less than 8 bytes, the minimum UDP packet size
+    #[error("Passed in data is not large enough for UDP header")]
+    MissingHeader,
+}
+
+/// UDP header and associated data
+pub struct UdpPacket {
```
Comment (johnyburd) [inline src/ip.rs]: maybe more correct to call it a datagram? since I think the whole thing is an ip packet and this is just a fragment of it

**Review by johnyburd — APPROVED** [review]

File: src/replay.rs (hunk 2 above)
Comment (jammiess) [reply src/replay.rs:93]: That is kinda just an example class. I guess I can add one for UDP as well

**Pull request merged**